hack/build.sh

```diff
@@ -30,6 +30,46 @@ copy_terraform_to_mirror() {
   cp "${PWD}/terraform/bin/${TARGET_OS_ARCH}/terraform" "${PWD}/pkg/terraform/providers/mirror/terraform/"
 }
 
+# Check if a provider has changed based on the version stored in the binary
+check_module_changes() {
+	binpath="$1"
+	srcpath="$2"
+	build_hash="$(go version -m "${binpath}" | grep 'vcs.revision' | cut -f2 -d'=')"
+	# If it's a locally developed provider
+	if test -n "${build_hash}"; then
+		# Check if a provider has changed based on git changes since the
+		# revision the provider was last built
+		git diff --name-only --exit-code "${build_hash}.." -- "${srcpath}"
+	# If it's an imported module
+	else
+		# Check if a provider has changed based on its go.mod git hash
+		version_info="$(go version -m "${binpath}" | grep -Eo 'installer.version=[a-fA-F0-9]+' | cut -f2 -d'=')"
+		test "${version_info}" == "$(git hash-object "${srcpath}/go.mod")"
+	fi
+}
+
+# Build terraform and providers only if needed
+build_terraform_and_providers() {
+	TARGET_OS_ARCH=$(go env GOOS)_$(go env GOARCH)
+	bindir="${PWD}/terraform/bin/${TARGET_OS_ARCH}"
+	find "${PWD}/terraform/providers/" -maxdepth 1 -mindepth 1 -type d | while read -r dir; do
+		provider="$(basename "${dir}")"
+		binpath="${bindir}/terraform-provider-${provider}"
+		if test -s "${binpath}" && test -s "${binpath}.zip" && check_module_changes "${binpath}" "terraform/providers/${provider}" ]; then
```

Comment:
```suggestion
		if test -s "${binpath}" && test -s "${binpath}.zip" && check_module_changes "${binpath}" "terraform/providers/${provider}"; then
```
I'll fix this in the next rebase.